src/common/src/row/mod.rs

```diff
@@ -12,21 +12,341 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-pub mod vec_datum;
+mod compacted_row;
+mod owned_row;
 
-pub use vec_datum::{Row, RowDeserializer};
+use std::hash::{BuildHasher, Hasher};
 
-/// `CompactedRow` is used in streaming executors' cache, which takes less memory than `Vec<Datum>`.
-/// Executors need to serialize Row into `CompactedRow` before writing into cache.
-#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
-pub struct CompactedRow {
-    pub row: Vec<u8>,
+pub use compacted_row::CompactedRow;
+pub use owned_row::{Row, RowDeserializer};
+
+use crate::array::RowRef;
+use crate::hash::HashCode;
+use crate::types::{hash_datum_ref, to_datum_ref, Datum, DatumRef, ToDatumRef, ToOwnedDatum};
+use crate::util::value_encoding;
+
+pub trait Row2: Sized + std::fmt::Debug + PartialEq + Eq {
+    type Iter<'a>: Iterator<Item = DatumRef<'a>>
+    where
+        Self: 'a;
+
+    fn datum_at(&self, index: usize) -> DatumRef<'_>;
+
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn iter(&self) -> Self::Iter<'_>;
+
+    fn to_owned_row(&self) -> Row {
+        Row(self.iter().map(|d| d.to_owned_datum()).collect())
+    }
+
+    fn into_owned_row(self) -> Row {
+        self.to_owned_row()
+    }
+
+    fn value_serialize(&self) -> Vec<u8> {
```

Comment:
Is there any specific reason to call it `value_serialize`? Maybe `value_encoding` or serialize is more clear and succinct.

Reply:
I name it `serialize` to follow the convention of Rust libraries. And `value` is to show that it should be used in the value of kv store. We may add a `key_serialize` or `ordered_serialize` in the future as well.